R/driver-sql-server.R

```diff
@@ -204,3 +204,56 @@ setMethod("odbcDataType", "Microsoft SQL Server",
     )
   }
 )
+
+#' @rdname SQLServer
+#' @description
+#' ## `odbcListObjects()`
+#'
+#' This method makes tables that are synonyms visible in the Connections pane.
+# See (#221).
+#' @usage NULL
+#' @export
+`odbcListObjects.Microsoft SQL Server` <- function(connection,
+                                                   catalog = NULL,
+                                                   schema = NULL,
+                                                   name = NULL,
+                                                   type = NULL,
+                                                   ...) {
+  objects <- NextMethod(
+    object = connection,
+    catalog = catalog,
+    schema = schema,
+    name = name,
+    type = type,
+    ...
+  )
+
+  if (is.null(catalog) | is.null(schema)) {
+    return(objects)
+  }
+
+  synonyms <- dbGetQuery(
+    connection,
+    "SELECT
+       catalog = DB.name,
+       [schema] = Sch.name,
+       name   = Syn.name
+     FROM sys.synonyms          AS Syn
+       INNER JOIN sys.schemas AS Sch
+         ON Sch.schema_id = Syn.schema_id
+       INNER JOIN sys.databases AS DB
+         ON Sch.principal_id = DB.database_id
+     WHERE DB.name = ? AND Sch.name = ?
+       AND OBJECTPROPERTY(Object_ID(Syn.base_object_name), 'IsTable') = 1;",
```

Comment:
For now, I've opted to only include synonyms whose base object is a table, and refer to that synonym as a "table" in the output. This means that synonyms (that are tables) will be previewable through the Connections pane, but also feels a bit hacky. From what I can tell, the drop-down based in the pane based on `odbcListFields()` also doesn't work (but doesn't cause errors/crashes).

Instead, we could opt not to include this table and return the `type` as `synonym`. This means that there would be entries in the Connections pane but no previews.